### ISSUE TITLE
DMA out getting stuck, PI4

### ISSUE THREAD

**Issue by kallekaninin** [mill1000/raspdif#8]
Hi.
Been trying to get this working as it seems like an excellent idea. 
Running a PI4, latest raspian lite.
I get no actual output on the pin, looking with a scope it is completely flat.
The program starts up and goes into "transmitting" but does not transmit anything.
Program feeding either via fifo or pipe will either just sit there och error out as the buffer never gets consumed.

I've narrowed it down to the following.

    if (dmaGetControlBlock(raspdif.dmaChannel) == &raspdif.control.bus->controlBlocks[buffer_index])
    {
      // If DMA is using current buffer, delay by approx 1 buffer's duration
      microsleep(1e6 * (RASPDIF_BUFFER_SIZE / arguments.sample_rate));
      continue;
    }

This block is executed every loop so looking at the comment I would assume that the dma-channel never gets emptied.

Also, this is the startup, the addresses looks right for a PI4.

D: Memory: Mapped physical address 0xFE000000 to virtual address 0xB530C000
D: Memory: Allocated memory at 0xFEBAF000 of length 98400.
D: Memory: Mapped physical address 0x3EBAF000 to virtual address 0xB52F3000
D: MAIN: Calculated SPDIF clock of 5.6448e+06 Hz for sample rate of 44100 Hz.
D: MAIN: Calculated DIVI: 88, DIVF: 2364.
I: MAIN: Estimated latency: 0.0928798 seconds.

Thanks.

**Comment by mill1000** [comment]
Hi. The Pi4 uses a different SoC from the previous Pi version so there is probably some compatibility issue. I'll look more in depth later but a quick search turned up 2 things.

- Try changing the DMA channel. I originally selected dma_channel_13 but [this](https://iosoft.blog/2020/05/25/raspberry-pi-dma-programming/) blog indicates it is being used by the system on Pi4.  Try switching it to channel 8, 9 or 10.
https://github.com/mill1000/raspdif/blob/bac2a4a44a64d5393ff9810a1cc87edeeaac2aa2/source/main.c#L422
- The Pi 4 has a larger memory space, it might be that the allocated memory is inaccessible to the DMA controller. There is talk of 40-bit addresses I'll have to investigate.

**Comment by kallekaninin** [comment]
Thanks. That was it.
The blog you linked has to be the best explanation of dma on the pi I've seen.
And as it suggested that dma 5 should be free on everything i tried that and it works fine!

Running a preexisting file via ffmpeg either trough fifo or a pipe works perfectly.
Doing it with something thats generated on the fly does not.
Buffer underrun all the time.
Either hook straight into alsa or pipe something into ffmpeg then out to raspdif.
Will try messing around with the buffering from source to ffmpeg but this might not be solvable without turning raspdif into a kernel driver as a real soundcard.
Or delaying the audio a lot to make sure the buffer stays full.

Btw. I think that in case a buffer underrun does occur it might be a good idea to pad the output with silence as there is an annoying pop whenever the output starts up again (might depend on receiver).

**Comment by mill1000** [comment]
Excellent! Glad to hear you've got it working to some degree.

What are you using to generate data "on the fly"? I've used this with both gmrender-resurrect and shairport-sync without issue to stream from other devices. I'm not sure if that counts as "on the fly". The input buffer is **really** primitive. It will declare an buffer underrun as soon as the pipe or file would block, even though there may be remaining data in the buffers. I guess you could call this a known issue :)

As of commit 38115bdb9987753cc36c0890376cdcb5e1f64f74 I switched to filling the output with silence instead of shutting off the output. It reduces the number of pops on my receiver quite a bit but they still appear every now and then between track transitions. If you aren't using the latest commit I'd recommend you try that.

**Comment by kallekaninin** [comment]
What I'm actually trying to do is run a synth.
I've had the software write to /tmp/out, pipe that into ffmpeg and out to raspdif.
(Only get buffer underrun after a second or so)

Which strangely enough works slightly better than the following which is abusing alsa a bit.
But if it is possible to get that to work then anything playing sound will work..
Load modprobe snd-aloop
Rec: ffmpeg -f alsa -i hw:0 -f s16le -acodec pcm_s16le - | sudo raspdif
Send: speaker-test -D hw:Loopback,0,0 -c 2
(Or something like that, addresses might be a bit off. Only works if no other sound card is loaded)
It works but there is constant underruns.

Also just tried gstreamer. Playing an existing flac directly works fine. Doing it from the fifo out from the synth does not. Exact same behavior.
Have not been able to connect gstreamer to the correct alsa output. Am going to try loading an actual souncard and cloning that instead since i guess things might get a bit confused.

**Comment by mill1000** [comment]
Interesting.  Just spit balling here. Does it work any better if you have ffmpeg write to a fifo, and have raspdif read from that?

Another thought. You could change the condition here to not zero-fill and poll the input. Instead you could just set the samples to 0 when the read fails. Your receiver should always see a valid stream of data, the downside is raspdif will continue to use CPU time even when there is no data.
https://github.com/mill1000/raspdif/blob/7e7679de75b9871f85ca9804c5c8c8d5ebcbeeb5/source/main.c#L483

**Comment by kallekaninin** [comment]
Checked my command history and i did try having ffmpeg do like that and it behaves the same.

I tried your idea and this gets me very close.
I had to make a ugly modification in the loop that get 2 samples at a time and only take 1 and increment a counter and continue the loop unless there are 2 complete samples.
Playing the old prerecorded file works just as well as before. The live audio is almost completely there.
Also this time i played the same file with ffmpeg set to output to alsa so that i could compare.

This is hooked on the output of alsa via ffmpeg to massage the data.
But of course raspdif now steals 50% of the cpu so this might not exactly be optimal.
Im getting a "pop" at about 3 times/sec, very short. 
See the following ffmpeg output. This looks alright?

```
Guessed Channel Layout for Input Stream #0.0 : stereo
Input #0, alsa, from 'hw:Loopback,1,0':
  Duration: N/A, start: 1608228683.411749, bitrate: 1536 kb/s
    Stream #0:0: Audio: pcm_s16le, 48000 Hz, stereo, s16, 1536 kb/s
File '/tmp/spdif' already exists. Overwrite ? [y/N] y
Stream mapping:
  Stream #0:0 -> #0:0 (pcm_s16le (native) -> pcm_s16le (native))
Press [q] to stop, [?] for help
Output #0, s16le, to '/tmp/spdif':
  Metadata:
    encoder         : Lavf58.20.100
    Stream #0:0: Audio: pcm_s16le, 48000 Hz, stereo, s16, 1536 kb/s
    Metadata:
      encoder         : Lavc58.35.100 pcm_s16le
size=   84263kB time=00:07:29.41 bitrate=1536.0kbits/s speed=1.01x    
```



EDIT: Yes raspdif is now running at 48khz.

**Comment by mill1000** [comment]
There must be another underlying incompatibility. I was able to run the loopback test you mentioned earlier without any underruns on RPi3
```
$ ffmpeg -f alsa -i hw:0 -f s16le -acodec pcm_s16le - | sudo raspdif  -v -r 48000
```
```
$ speaker-test  -c 2 -D hw:0,1
```

EDIT:

Double check that the input and output are using the same rate.
e.g.
```
$ ffmpeg -f alsa -i hw:0 -f s16le -acodec pcm_s16le - | sudo raspdif  -v -r 48000
```
will stutter if I try the following
```
speaker-test  -c 2 -D hw:0,1 -r 44100
```

**Comment by kallekaninin** [comment]
Made a few more tests to make sure i didnt miss anything.
Fresh copy of your repo, only tried both dma 5 and 10 and get this result directly.

```
sudo modprobe snd-aloop
ffmpeg -f alsa -i hw:0 -f s16le -acodec pcm_s16le - | sudo raspdif  -v -r 48000
D: Memory: Mapped physical address 0xFE000000 to virtual address 0xB536C000
D: Memory: Allocated memory at 0xFEBAF000 of length 98400.
D: Memory: Mapped physical address 0x3EBAF000 to virtual address 0xB5353000
D: MAIN: Calculated SPDIF clock of 6.144e+06 Hz for sample rate of 48000 Hz.
D: MAIN: Calculated DIVI: 81, DIVF: 1557.
I: MAIN: Estimated latency: 0.0853333 seconds.
I: MAIN: Waiting for data...
ffmpeg version 4.1.6-1~deb10u1+rpt1 Copyright (c) 2000-2020 the FFmpeg developers
--------snip--------
Guessed Channel Layout for Input Stream #0.0 : stereo
Input #0, alsa, from 'hw:0':
  Duration: N/A, start: 1608234354.564589, bitrate: 1536 kb/s
    Stream #0:0: Audio: pcm_s16le, 48000 Hz, stereo, s16, 1536 kb/s
Stream mapping:
  Stream #0:0 -> #0:0 (pcm_s16le (native) -> pcm_s16le (native))
Press [q] to stop, [?] for help
Output #0, s16le, to 'pipe:':
  Metadata:
    encoder         : Lavf58.20.100
    Stream #0:0: Audio: pcm_s16le, 48000 Hz, stereo, s16, 1536 kb/s
    Metadata:
      encoder         : Lavc58.35.100 pcm_s16le
I: MAIN: Transmitting...
D: MAIN: Buffer underrun.
D: MAIN: Data available.
D: MAIN: Buffer underrun.
D: MAIN: Data available.
D: MAIN: Buffer underrun.0:00.51 bitrate=1521.4kbits/s speed=1.02x    
D: MAIN: Data available.
D: MAIN: Buffer underrun.
D: MAIN: Data available.
D: MAIN: Buffer underrun.0:01.01 bitrate=1526.5kbits/s speed=1.01x    
D: MAIN: Data available.
-----------------------------------------
```

I'll continue messing with it tomorrow.

**Comment by kallekaninin** [comment]
Played around a bit with things today. Unfortunately my pi3 suffers from being connected to the wrong voltage and is resting in the scrapheap so I cant really compare.
After messing around and cutting out ffmpeg altogether, just alsa writing to a fifo and raspdif receiving (which works just as fine since alsa resamples anything incoming for proper mixing). Got a few support scripts and things to make this painless if you would like to see if this works on your pi3 as well (anything playing to default will be dumped to the fifo and soundcard).

Anyway; Got to thinking a bit, sending at 48khz and rec at 48khz, still gaps in the audio that would very much happen if there was a missmatch.
So, the clocksource. Which led me down a rabbit-hole about clock synchronization in audio applications and the general consensus seem to be that it's Difficult(tm).

This would explain a lot since when a prerecorded file is played the pace is pretty much set by the output. So the playback might not be exactly 1x. The difference would be extremely small, so unlikely to be noticed by ear.
The live source on the other hand has to work the opposite since the source is by definition stuck at 1x.

I've just spent some time taking a look on alsa interfaces. Turns out it seems fairly easy to capture audio using those interfaces. They even support mmap to make it a bit more efficient.
So I'm going to throw together a version that does exactly that and thus letting alsa set the pace. Might introduce a bit of jitter in the spdif output but when i mess with at least my receiver it seems fairly tolerant.

What do you think? I could very well be on the wrong track, it has been a long week.

**Comment by mill1000** [comment]
I'd be interested in trying your scripts. Having ALSA send to the fifo seems like a great solution for general purpose use. Would make raspdif usable for applications that can't write to a file or pipe.

I have a spare Pi3 that I can load with the latest Raspbian. I think my dev board is technically running on Raspbian Stretch, so it's possible the issue is due to the newer OS and not the Pi4 hardware.

**Comment by kallekaninin** [comment]
Hi.
Didn't mean to ghost you. Simply this fxxin year..
Dug out the Pi that I used for testing this and the sd-card is gone. And with it so are my changes.
So this is from memory.
I believe this is the example i used for making alsa split the streams. Change the output rate and file to match.
https://www.raspberrypi.org/forums/viewtopic.php?t=251299
then a simple systemd service
example
https://stackoverflow.com/questions/48195340/systemd-with-multiple-execstart
that did mkfifo FILE && chmod 777 FILE
Maybe set proper permissions?
There are at least one alsa systemd unit file, two on my pi. Change them so that the After= is the new mkfifo service.
Move anything that the alsa services depended on into the mkfifo service

There are presumably a proper way to inject a service between two others but i could not be bothered.

I then simply ran raspdif manually since i was still trying to find the problems.
You could of course put that in the script also.

**Comment by mill1000** [comment]
No worries. Thanks for those links. I'll have to try that setup sometime.

I can't remember exactly where we left off. Was there a remaining issue we were trying to solve?

**Comment by kallekaninin** [comment]
Well, the sound works as long as i run something prerecorded. If I try to use "live" audio in any way (including the alsa-hack above) there will be many many buffer underruns.
This was with a pi4.
Feel free to drop me a line if you want me to test something.

**Comment by mill1000** [comment]
@kallekaninin I've made a few changes to how the FIFO is read is this branch: https://github.com/mill1000/raspdif/tree/poll_fd

It will now block for a short amount of time before declaring an underrun. Maybe this will help with your issue.

**Comment by ajeddeloh** [comment]
I have a different rpi 4 issue: my clock seems to run at ~~3/4~~ 1.5x speed. I've confirmed this on my oscilloscope. I can workaround by just setting the sample rate ~~4/3 higher~~ 2/3 slower (i.e. ~~64k~~ 32k instead of 48k). I've tried with both 44.1k and 48k on dma channels 8, 9 and 10. This is running with /dev/zero as the input file:

```
pi@raspberrypi:~/raspdif $ sudo ./build/raspdif -v -r 48000 -i /dev/zero -f s16le
D: Memory: Mapped physical address 0xFE000000 to virtual address 0xB52F8000
D: Memory: Allocated memory at 0xFEBE2000 of length 98400.
D: Memory: Mapped physical address 0x3EBE2000 to virtual address 0xB52DF000
D: MAIN: Calculated SPDIF clock of 6.144e+06 Hz for sample rate of 48000 Hz.
D: MAIN: Calculated DIVI: 81, DIVF: 1557.
I: MAIN: Estimated latency: 0.0853333 seconds.
I: MAIN: Waiting for data...
I: MAIN: Transmitting...
```

EDIT: got confused by the encoding, thought it was too slow, was actually too fast.

**Comment by mill1000** [comment]
The clock dividers are calculated assuming a 500 MHz PLLD frequency. All the document ion I can find seems to indicate this frequency is fixed, however yours may be running faster.

What's the output of this command on your system? `sudo cat /sys/kernel/debug/clk/clk_summary | grep plld`

e.g. On mine
```
    plld                              3        3        0  2000000024          0     0  50000
       plld_dsi1                      0        0        0     7812501          0     0  50000
       plld_dsi0                      0        0        0     7812501          0     0  50000
       plld_per                       2        2        0   500000006          0     0  50000
       plld_core                      2        2        0   500000006          0     0  50000
```

**Comment by ajeddeloh** [comment]
Yup I think that's the issue

```
    plld                              5        5        0  3000000091          0     0  50000
       plld_dsi1                      1        1        0    11718751          0     0  50000
       plld_dsi0                      1        1        0    11718751          0     0  50000
       plld_per                       3        3        0   750000023          0     0  50000
       plld_core                      1        1        0   600000019          0     0  50000
```

**Comment by mill1000** [comment]
It appears the Pi4 does run the PLLD at 750 MHz: https://www.raspberrypi.org/forums/viewtopic.php?f=72&t=244479&start=25

Quick fix for you is to change the value here:
https://github.com/mill1000/raspdif/blob/facd46e38abe6084cd194fae94ed2e540644a85b/source/main.c#L199

**Comment by mill1000** [comment]
Closed via c0d93de
